labelify/__main__.py

```diff
@@ -188,28 +192,134 @@ def file_or_dir_path(path):
         default="false",
     )
 
+    parser.add_argument(
+        "-u",
+        "--username",
+        type=str,
+        default=None,
+        dest="username",
+        help="sparql username",
+        required=False,
+    )
+
+    parser.add_argument(
+        "-p",
+        "--password",
+        type=str,
+        default=None,
+        dest="password",
+        help="sparql password",
+        required=False,
+    )
+
+    parser.add_argument(
+        "-g",
+        "--graph",
+        type=str,
+        default=None,
+        dest="graph",
+        help="named graph to query (only used if input is a sparql endpoint)",
+        required=False,
+    )
+
+    parser.add_argument(
+        "-r",
+        "--raw",
+        dest="raw",
+        action="store_true",
+        help="Only output the nodes with missing labels. one per line",
+    )
+
     args = parser.parse_args(args)
 
-    g = Graph().parse(args.input)
+    if isinstance(args.input, ParseResult):
+        g = Graph()
+        offset = 0
+        batch_size = 50000
+        n_results = batch_size
+        url = urlunparse(args.input)
+        if not args.raw:
+            print(
+                f"Loading triples from {url}, using graph: {'default' if not args.graph else args.graph}"
+            )
+            print(f"\tbatch_size: {batch_size}")
+        while n_results == batch_size:
+            sparql = SPARQLWrapper(endpoint=url, defaultGraph=args.graph)
+            sparql.setReturnFormat(JSONLD)
+            sparql.setQuery(
+                """
+            construct {{
+                ?s ?p ?o .
+            }}
+            where {{
+                ?s ?p ?o .
+            }}
+            order by ?s
+            limit {}
+            offset {}
+            """.format(
+                    batch_size, offset
+                )
+            )
+            if args.username and not args.password:
+                sparql.setCredentials(user=args.username, passwd=getpass("password:"))
+                if not args.raw:
+                    print("\n")
+            elif args.username and args.password:
+                sparql.setCredentials(user=args.username, passwd=args.password)
```

Comment:
Do you reckon we can move the instantiation of the SPARQLWrapper object outside of the while loop? This way we should be able to speed up the code a bit by reducing the need to create a new object for each iteration. Each iteration, we should be able to just do `sparql.setQuery` and then call `sparql.queryAndConvert`.

If sparqlwrapper allows for connection pooling, we should also use that to avoid creating a new tcp connection on each iteration. My feeling is that it probably doesn't support this :(

Are you also able to set a default timeout by calling `sparql.setTimeout()`? Would be good to have it timeout if the server decides to hang. We may want to provide another CLI flag for users to set their custom timeout value.

Reply:
Thanks. I've applied those updates. appreciate the thorough review.